# Segmenting and Clustering Neighborhoods in Toronto

## Part 1: Importing Neighbourhood Data from Wikipedia

In this section, I scrape the wikipedia page 'List of postal codes of Canada: M' (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) to get Toronto postal code, borough, and neighbourhood data into a Pandas dataframe.

First, need to import pandas:

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

The wikipedia page contains several tables. We are only interested in the first one, which I assign to a pandas dataframe and preview the dataframe:

In [2]:
wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(wiki, header=0)[0]
df.head(10)

Postal Code           Borough                                 Neighborhood
0         M1A      Not assigned                                          NaN
1         M2A      Not assigned                                          NaN
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                          NaN
8         M9A         Etobicoke                             Islington Avenue
9         M1B       Scarborough                               Malvern, Rouge

We are only interested in processing cells that have an assigned borough. Removing rows with unassigned boroughs:

In [3]:
df = df[df.Borough != 'Not assigned']
df.head(10)

Postal Code           Borough                                 Neighborhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke                             Islington Avenue
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson

Number of rows in the dataframe: 

In [4]:
print('The number of rows in the dataframe is', df.shape[0])

The number of rows in the dataframe is 103


Therefore we are dealing with 103 neighbourhoods in Toronto.

## Part 2: Importing Latitude and Longitude Data

Importing the latitude and longitude data for each neighbourhood into a pandas dataframe from csv file: http://cocl.us/Geospatial_data:

In [5]:
geo_url = 'http://cocl.us/Geospatial_data'
geo_df = pd.read_csv(geo_url)
geo_df.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

Set indices to 'Postal Code' for both dataframes to allow for merging:

In [6]:
geo_df = geo_df.set_index('Postal Code')
df = df.set_index('Postal Code')

Concatenate the two dataframes into a new dataframe:

In [7]:
all_df = pd.concat([df, geo_df], axis=1, join='outer', sort=False)
all_df.head(10)

Borough                                 Neighborhood   Latitude  \
M3A        North York                                    Parkwoods  43.753259   
M4A        North York                             Victoria Village  43.725882   
M5A  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
M6A        North York             Lawrence Manor, Lawrence Heights  43.718518   
M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
M9A         Etobicoke                             Islington Avenue  43.667856   
M1B       Scarborough                               Malvern, Rouge  43.806686   
M3B        North York                                    Don Mills  43.745906   
M4B         East York              Parkview Hill, Woodbine Gardens  43.706397   
M5B  Downtown Toronto                     Garden District, Ryerson  43.657162   

     Longitude  
M3A -79.329656  
M4A -79.315572  
M5A -79.360636  
M6A -79.464763  
M7A -79.389494  
M9A -79.532242  
M1B -79.194353  
M3B -79.352188  
M4B -79.309937  
M5B -79.378937

Reset index of the new dataframe and recover 'Postal Code' column title:

In [8]:
all_df.reset_index(inplace=True)
all_df.rename(columns={'index': 'Postal Code'},inplace=True)
all_df.head(10)

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke                             Islington Avenue   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

## Part 3: Clustering Toronto Neighbourhoods

For the purpose of clutering, I will focus only on boroughs that contain the word 'Toronto'. Therefore I will create a new dataframe containing  neighbourhoods corresponding only to these boroughs:

In [24]:
neighborhoods = all_df[all_df['Borough'].astype(str).str.contains('Toronto')]
neighborhoods

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68                            Forest Hill North & West  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                                  North Toronto West  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

Importing relevant dependencies:

In [17]:
import numpy as np # library to handle data in a vectorized manner

#!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!pip install folium==0.5.0
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Obtaining latitude and longitude coordinates of Toronto for mapping:

In [18]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Creating a map of Toronto with neighborhoods superimposed:

In [25]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Clustering Neighborhoods by Borough

Cluster the neighbourhoods into boroughs, creating separate dataframes for each borough:

In [53]:
downtown_toronto = neighborhoods[neighborhoods['Borough']=='Downtown Toronto']
central_toronto = neighborhoods[neighborhoods['Borough']=='Central Toronto']
west_toronto = neighborhoods[neighborhoods['Borough']=='West Toronto']
east_toronto = neighborhoods[neighborhoods['Borough']=='East Toronto']

#### Creating a map of Toronto with neighborhoods superimposed, coloured by borough:

In [58]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add dt Toronto markers to map
for lat, lng, borough, neighborhood in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], downtown_toronto['Borough'], downtown_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
for lat, lng, borough, neighborhood in zip(central_toronto['Latitude'], central_toronto['Longitude'], central_toronto['Borough'], central_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

for lat, lng, borough, neighborhood in zip(west_toronto['Latitude'], west_toronto['Longitude'], west_toronto['Borough'], west_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='green',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
for lat, lng, borough, neighborhood in zip(east_toronto['Latitude'], east_toronto['Longitude'], east_toronto['Borough'], east_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='purple',
        fill=True,
        fill_color='purple',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

In [ ]:
#kclusters = 5
#k_means = KMeans(init="k-means++", n_clusters=kclusters, n_init=12)
#X = neighborhoods.values[:,4:6]
#boroughs = pd.get_dummies(neighborhoods[['Borough']], prefix="", prefix_sep="")